Address comments in PR Snowbridge-v2

yrong opened the pull request.

yrong committed:
Move to the end

yrong committed:
Rename as RemoteExecutionFee

yrong committed:
Document the config type

yrong committed:
Introduce XcmExecutor and charge delivery fees

yrong committed:
Improve comments

yrong committed:
Remove comment

yrong committed:
Versioning for ContractCall and improve documentation

yrong committed:
More comments

yrong committed:
Rename as DeliveryReceiptDecodeError

yrong committed:
Rename file to delivery_receipt.rs

yrong committed:
Improve docs

yrong committed:
Delete the readme

yrong committed:
Rename to DeliveryReceipt

yrong committed:
More comments

yrong committed:
Remove second_governance_origin

yrong committed:
Factor this ENA&PNA block into a separate function

yrong committed:
Pass bridgehub location via config

yrong committed:
Fix tests

yrong committed:
Remove root governance on BH

yrong committed:
Improve comment

yrong committed:
Move type config to the end

yrong committed:
Rename to EthereumSystemCall

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -198,16 +205,18 @@ pub mod pallet {
 		}
 		pub fn burn_fees(origin_location: Location, fee: u128) -> DispatchResult {
 			let ethereum_fee_asset = (T::FeeAsset::get(), fee).into();
-			T::AssetTransactor::withdraw_asset(&ethereum_fee_asset, &origin_location, None)
+			T::XcmExecutor::charge_fees(origin_location.clone(), ethereum_fee_asset)
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
-			T::AssetTransactor::withdraw_asset(&T::DeliveryFee::get(), &origin_location, None)
+			T::XcmExecutor::charge_fees(origin_location, T::RemoteExecutionFee::get().into())
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
 			Ok(())
 		}
```
This whole function is problematic.

Firstly, you cannot initiate a teleport (burning) by using `T::XcmExecutor::charge_fees`. That does not burn, it simply transfers the value to a treasury account. Rather use `T::AssetTransactor::withdraw_asset` like you were doing.

Secondly, the function is confusing, because its signature is `burn_fees(location, fee)`, which implies that only `fee` is being being burnt. However the function is also secretly burning `T::RemoteExecutionFee::get()`, which is confusing for reviewers.

Rather create a generic `burn_for_teleport(location, asset)` function that is called twice:

```rust
burn_for_teleport(origin_location, ether_fee);
burn_for_teleport(origin_location, remote_execution_fee);
```

yrong replied:
Nice catch! 
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/c9369a615cb7f75bc50177db1b0fc12048b425e5 for the fix.

Btw: For less confusion I've removed the  `burn_fees` function. Extracting `burn_for_teleport` also seems unnecessary since it contains only a single line of code, so I prefer to inline it instead.

vgeddes replied:
I prefer factoring out the burning into `burn_for_teleport`.

Also you must also call `AssetTransactor::can_check_out`, like what we had to do for V1.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/af8205340d5860f51e3dc49606471320084fc96b

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -75,17 +74,23 @@ pub mod pallet {
 		type XcmSender: SendXcm;
 
-		/// To withdraw and deposit an asset.
-		type AssetTransactor: TransactAsset;
```
No, you still need to use this for burning assets.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -75,17 +74,23 @@ pub mod pallet {
-
-		#[cfg(feature = "runtime-benchmarks")]
-		type Helper: BenchmarkHelper<Self::RuntimeOrigin>;
+		/// Handler for XCM fees.
```
```suggestion
		/// Something to charge XCM delivery fees
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -75,17 +74,23 @@ pub mod pallet {
+		type RemoteExecutionFee: Get<Asset>;
+
+		/// Location of bridge hub
+		type BridgeHub: Get<Location>;
```
```suggestion
		type BridgeHubLocation: Get<Location>;
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/6bbec7ddd0b10161ca2a26214df0952be606283b

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -198,16 +205,18 @@ pub mod pallet {
-			send_xcm::<T::XcmSender>(bridgehub, xcm).map_err(|_| Error::<T>::Send)?;
+		pub fn send(origin: Location, xcm: Xcm<()>) -> DispatchResult {
+			let bridgehub = T::BridgeHub::get();
+			let (_, price) =
```
You should also include the XCM message ID in the event, for tracing purposes

```suggestion
			let (message_id, price) =
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/282208d99fd8c84ef695bd258c7d05d4156c4f69

Btw:  I assume the send hash is not the `message_id` for tracing purposes. To achieve that, we need a `SetTopic` instruction.

vgeddes replied:
Nope. if the XCM message ends with a trailing topic, that topic will be used as the message_id. See

 See https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/cumulus/parachains/runtimes/bridge-hubs/bridge-hub-westend/src/xcm_config.rs#L242

https://paritytech.github.io/polkadot-sdk/master/staging_xcm_builder/struct.WithUniqueTopic.html

yrong replied:
Yeah, with the `WithUniqueTopic` router the send hash happens to be same as the `topic_id`,  but the key is  `SetTopic` instruction. On receiver end with the barrier [TrailingSetTopicAsId](https://github.com/paritytech/polkadot-sdk/blob/f340e07ffd7cff69e930d19563186ff7f0fed601/cumulus/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs#L282) the `topic_id` is respected.

That’s basically how the indexer works under the hood.

yrong replied:
Anyway, the send hash seems easier to understand. Changed in 
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/004c84b4cb2f3e7c0d5dc50c425ba8bd91796680

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -189,7 +196,7 @@ pub mod pallet {
-			Self::send(xcm)?;
+			Self::send(origin_location, xcm)?;
 
 			Self::deposit_event(Event::<T>::RegisterToken { location: asset_location });
```
Include the message_id in the event.

vgeddes commented on bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs:
```diff
@@ -109,18 +111,103 @@ where
-	fn to_ethereum_message(&mut self) -> Result<Message, XcmConverterError> {
-		use XcmConverterError::*;
+	/// Extract ethereum native assets
+	fn extract_enas(
```
```suggestion
	fn extract_ethereum_native_assets(
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/66c2400b6c7f1f8732b720936bb6c58a5929446c

vgeddes commented on bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs:
```diff
@@ -109,18 +111,103 @@ where
+	}
+
+	/// Extract PNA
+	fn extract_pnas(
```
```suggestion
	fn extract_polkadot_native_assets(
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/66c2400b6c7f1f8732b720936bb6c58a5929446c

vgeddes commented on bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs:
```diff
@@ -100,7 +99,7 @@ pub use types::{PendingOrder, ProcessMessageOriginOf};
 use xcm::latest::{Location, NetworkId};
 
-type MessageReceiptOf<T> = MessageReceipt<<T as frame_system::Config>::AccountId>;
+type MessageReceiptOf<T> = DeliveryReceipt<<T as frame_system::Config>::AccountId>;
```
```suggestion
type DeliveryReceiptOf<T> = DeliveryReceipt<<T as frame_system::Config>::AccountId>;
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/8849adf6feb45653552f53616ecd455f9d1a5b79

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -198,16 +205,18 @@ pub mod pallet {
+				send_xcm::<T::XcmSender>(bridgehub, xcm).map_err(|_| Error::<T>::Send)?;
+			T::XcmExecutor::charge_fees(origin, price).map_err(|_| Error::<T>::FundsUnavailable)?;
 			Ok(())
 		}
 		pub fn burn_fees(origin_location: Location, fee: u128) -> DispatchResult {
```
```suggestion
		}

		pub fn burn_fees(origin_location: Location, fee: u128) -> DispatchResult {
```

yrong committed:
Revert "Introduce XcmExecutor and charge delivery fees"

This reverts commit 0735da97a850b59be20ff8122322234bd0281187.

yrong committed:
Rename to BridgeHubLocation

yrong committed:
Add message_id for tracing

yrong committed:
Rename function

yrong committed:
Rename to DeliveryReceiptOf

yrong committed:
Rename to InvalidAssetOwner

yrong committed:
Fix build

yrong committed:
Fix config

yrong committed:
Cleanup

yrong commented:
@vgeddes This PR is ready for review. To limt the scope and the review process easier,  I'll address other comments in https://github.com/paritytech/polkadot-sdk/pull/7402 in seperated PRs.

yrong committed:
Fix frontend calls -1

yrong committed:
Remove origin_location from Message

yrong committed:
Merge branch 'ron/snowbridge-v2-for-comments' into ron/snowbridge-v2-for-comments-fix-frontend-calls

yrong committed:
Ignore fee charge for sudo

yrong committed:
Remove prefix check for registering token

yrong committed:
Polish

yrong commented:
@vgeddes I've addressed most of the comments in the original PR. 

To limt the scope and make the review process easier,  I may not cover all the left in this one, instead add more seperate PRs on top of this one as follows:

- set a halt boolean flag in runtime storage and make a custom exporter to respect that flag as mentioned 
in https://github.com/paritytech/polkadot-sdk/pull/7402#discussion_r1948158299
- add governace calls in system-v2 pallet for `upgrade` and `set_operating_mode extrinsics` for https://github.com/paritytech/polkadot-sdk/pull/7402#discussion_r1948231637
- Remove the outbound runtime api

yrong committed:
Fix comments

yrong committed:
More tests

yrong committed:
Polish

yrong committed:
Remove primary_governance_origin

yrong committed:
Refactor burn_for_teleport

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,78 @@ pub mod pallet {
-		/// - `origin`: Can be any sovereign `Location`
+		/// - `origin`: `Location` can be from either of:
+		/// 	a. any kind of location from a sibling parachain
+		/// 	b. sudo from local, location will be `Here` in this case.
```
See my comment below, I think there is a high chance your assumption that `governance == Here` is incorrect.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -42,9 +44,9 @@ pub enum ControlCall {
+pub enum BridgeHubRuntime {
 	#[codec(index = 90)]
-	Control(ControlCall),
+	Control(EthereumSystemCall),
```
```suggestion
	EthereumSystem(EthereumSystemCall),
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/4d82f3b875149e607bc1b0764d29bb049a5bf6e1

vgeddes commented on bridges/snowbridge/pallets/outbound-queue-v2/src/send_message_impl.rs:
```diff
@@ -39,7 +42,10 @@ where
 	fn deliver(ticket: Self::Ticket) -> Result<H256, SendError> {
 		let origin = AggregateMessageOrigin::SnowbridgeV2(ticket.origin);
 
-		if ticket.origin != primary_governance_origin() {
+		let governance_origin =
+			AgentIdOf::convert_location(&Here.into()).ok_or(SendError::InvalidOrigin)?;
+
+		if ticket.origin != governance_origin {
 			ensure!(!Self::operating_mode().is_halted(), SendError::Halted);
 		}
```
This is all obsolete code, since the pallet will no longer have an operating mode. Please remove.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/6e1b0142618c433aece64e2798938874cead0b5c

vgeddes commented on bridges/snowbridge/pallets/outbound-queue-v2/src/mock.rs:
```diff
@@ -139,14 +141,17 @@ pub fn run_to_end_of_next_block() {
 	System::on_finalize(System::block_number());
 }
 
+pub fn primary_governance_origin() -> AgentId {
```
```suggestion
pub fn bridge_hub_root_origin() -> AgentId {
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,78 @@ pub mod pallet {
 
-			Self::burn_fees(origin_location.clone(), fee)?;
+			// Ignore fee charges for sudo call
+			if !origin_location.eq(&Here.into()) {
+				// Burn Ether Fee for the cost on ethereum
+				Self::burn_for_teleport(&origin_location, &(T::FeeAsset::get(), fee).into())?;
+
+				// Burn RemoteExecutionFee for the cost on bridge hub
+				Self::burn_for_teleport(&origin_location, &T::RemoteExecutionFee::get())?;
+			}
```
This just won't work (see problems below). Its really half baked. Please forget about `sudo` support for now, and get the permissionless version working and tested 100%.

1. You're teleporting the remote execution fee, and so it always needs to be burned locally.
2. Its not as simple as checking `location == Here` for governance. You need to use a `EnsureOrigin` checker configured by the Runtime.
3. We may want some snowbridge treasury account to fund fees for root, rather than skipping fees.

yrong replied:
All make sense!  Remove `sudo` in https://github.com/vgeddes/polkadot-sdk/pull/6/commits/e7c056ac98152a9956a07d51937e2eccb20e57b3

Btw: we've the permissionless version for:

[register token from asset hub](https://github.com/yrong/polkadot-sdk/blob/6e1b0142618c433aece64e2798938874cead0b5c/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs#L131)

[register token from penpal](https://github.com/yrong/polkadot-sdk/blob/6e1b0142618c433aece64e2798938874cead0b5c/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs#L419)

[register agent from penpal](https://github.com/yrong/polkadot-sdk/blob/6e1b0142618c433aece64e2798938874cead0b5c/cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs#L515)

vgeddes replied:
Thanks Ron.

I've additional tickets to cover the Rococo->Westend emulated tests for these functions: 
* https://linear.app/snowfork/issue/SNO-1339
* https://linear.app/snowfork/issue/SNO-1338

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/55bf251a5b2635a0e80a8377d3bdf8e11b76581c

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,78 @@ pub mod pallet {
 	impl<T: Config> Pallet<T> {
 		/// Call `create_agent` to instantiate a new agent contract representing `origin`.
-		/// - `origin`: Can be any sovereign `Location`
+		/// - `origin`: `Location` can be from either of:
```
We don't need to go into this amount of detail about the supported locations. Please shorten.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +207,76 @@ pub mod pallet {
-			Self::deposit_event(Event::<T>::RegisterToken { location: asset_location });
+			Self::deposit_event(Event::<T>::RegisterToken {
+				location: asset_location,
+				message_id: topic.into(),
```
See comment below, you should rather use the `message_id` returned by `send_xcm`.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,78 @@ pub mod pallet {
+		/// - `origin`: `Location` can be from either of:
+		/// 	a. any kind of location from a sibling parachain
+		/// 	b. sudo from local, location will be `Here` in this case.
+		/// 	c. a signed origin from local, location will be `(0,[AccountId32()])` in this case.
```
Too much detail, the runtime configuration could just override this, making it confusing.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/commit/e7c056ac98152a9956a07d51937e2eccb20e57b3

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +207,76 @@ pub mod pallet {
-			if asset_location.eq(&origin_location) || asset_location.starts_with(&origin_location) {
-				checked = true
+			// Ignore fee charges for sudo call
+			if !origin_location.eq(&Here.into()) {
```
Ditto, this `sudo` implementation has many problems. Please ignore `sudo` support for now.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +207,76 @@ pub mod pallet {
+			let (_, price) =
+				send_xcm::<T::XcmSender>(bridgehub, xcm).map_err(|_| Error::<T>::Send)?;
+			// Ignore fee charges for sudo call
+			if origin.clone() != Here.into() {
```
ignore sudo support for now please.

vgeddes commented:
> * add governace calls in system-v2 pallet for `upgrade` and `set_operating_mode extrinsics`

I think you will need to add them to `system-frontend` too, because for governance operations, the remote fees need to come from a treasury account, and those fees need to burnt on AH

But please do this in a separate PR as you stated

vgeddes commented on cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs:
```diff
@@ -19,10 +19,25 @@ use frame_support::pallet_prelude::TypeInfo;
 use xcm_executor::traits::ConvertLocation;
 
+#[derive(Encode, Decode, Debug, PartialEq, Clone, TypeInfo)]
+pub enum ControlFrontendCall {
```
```suggestion
pub enum EthereumSystemFrontendCall {
```

vgeddes commented on cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs:
```diff
@@ -19,10 +19,25 @@ use frame_support::pallet_prelude::TypeInfo;
+
+#[allow(clippy::large_enum_variant)]
+#[derive(Encode, Decode, Debug, PartialEq, Clone, TypeInfo)]
+pub enum SnowbridgeControlFrontend {
```
```suggestion
pub enum EthereumSystemFrontend {
```

vgeddes commented on cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs:
```diff
@@ -19,10 +19,25 @@ use frame_support::pallet_prelude::TypeInfo;
+#[derive(Encode, Decode, Debug, PartialEq, Clone, TypeInfo)]
+pub enum SnowbridgeControlFrontend {
+	#[codec(index = 80)]
+	Control(ControlFrontendCall),
```
```suggestion
	EthereumSystemFrontend(EthereumSystemFrontendCall),
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/4c1efaf74146c0038ffb9e52fd95cf6374cda30b

vgeddes commented on cumulus/parachains/integration-tests/emulated/tests/bridges/bridge-hub-westend/src/tests/snowbridge_v2_outbound.rs:
```diff
@@ -82,6 +97,68 @@ fn send_weth_from_asset_hub_to_ethereum() {
 }
 
+#[test]
+pub fn register_relay_token_from_asset_hub_with_sudo() {
```
Lets add tests for proper sudo support in a separate PR.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,78 @@ pub mod pallet {
 
-			let call = SnowbridgeControl::Control(ControlCall::CreateAgent {
-				location: Box::new(VersionedLocation::from(origin_location.clone())),
+			let reanchored_location = origin_location
```
I feel like BH should take responsibility for reanchoring. Does it not do that already?

vgeddes commented:
> > * add governace calls in system-v2 pallet for `upgrade` and `set_operating_mode extrinsics`
> 
> I think you will need to add them to `system-frontend` too, because for governance operations, the remote fees need to come from a treasury account, and those fees need to burnt on AH
> 
> But please do this in a separate PR as you stated

Added ticket to track this: https://linear.app/snowfork/issue/SNO-1336

yrong committed:
Remove sudo for frontend calls

yrong committed:
Remove sudo check

yrong committed:
Rename to EthereumSystem

yrong committed:
Remove origin check for governance

yrong committed:
Rename to EthereumSystemFrontendCall

yrong committed:
Rename to EthereumSystemFrontend

yrong committed:
Use message_id returned by send_xcm

yrong commented:
> add governance calls to `system-frontend` too

If that is a sudo call then by default the origin is mapped to `Here` as https://github.com/yrong/polkadot-sdk/blob/004c84b4cb2f3e7c0d5dc50c425ba8bd91796680/polkadot/xcm/xcm-builder/src/origin_conversion.rs#L250-L251 shows.

Then we may also need to custom the `LocationToAccountId` https://github.com/yrong/polkadot-sdk/blob/004c84b4cb2f3e7c0d5dc50c425ba8bd91796680/cumulus/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs#L92 mapping a location to an account from which fees are burned.

Just curious why you prefer to move governance calls to AH?  A fee-free call from root on BH is much easier and more straightforward, in my opinion. However, some minor changes may be required for the relayer to unconditionally relay messages from the governance origin.

Btw: by conventional, the root call is free in most cases, e.g.
https://github.com/paritytech/polkadot-sdk/blob/43ae08fa0783ff5974925b709d623d30990a2667/polkadot/xcm/pallet-xcm/src/lib.rs#L1810-L1817

yrong committed:
Remove dry run api

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,64 @@ pub mod pallet {
 				fee,
 			});
 
+			let topic = blake2_256(&("SnowbridgeFrontend", call.clone()).encode());
```
Actually, In realized this topic is not necessary. Since AH is configured to use `WithUniqueTopic`, one will be inserted automatically using hash of xcm. 

We should remove it.

yrong replied:
[Remove topic id](https://github.com/vgeddes/polkadot-sdk/pull/6/commits/a79df85275b33a695f2c296f1bf19f679564c52e)

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -78,25 +79,38 @@ pub mod pallet {
+		type XcmExecutor: ExecuteXcm<Self::RuntimeCall>;
 
+		/// Fee asset for the execution cost on ethereum
 		type FeeAsset: Get<Location>;
```
Strictly speaking, this a location, not a fee.
```suggestion
		/// Location of the remote ethereum network
		type EthereumLocation: Get<Location>;
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/ac5090e505ad9a41f931e0b7004b362d602d5782

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
-			Ok(())
+		pub fn send(origin: Location, xcm: Xcm<()>) -> Result<H256, Error<T>> {
+			let bridgehub = T::BridgeHubLocation::get();
+			let (hash, price) =
```
It doesn't have to be a hash. That's just an optional internal implementation detail.

```suggestion
			let (message_id, price) =
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
-			let bridgehub = Location::new(1, [Parachain(1002)]);
-			send_xcm::<T::XcmSender>(bridgehub, xcm).map_err(|_| Error::<T>::Send)?;
-			Ok(())
+		pub fn send(origin: Location, xcm: Xcm<()>) -> Result<H256, Error<T>> {
```
We can improve the error types returned here. Please see equivalent send function in inbound-queue-v2:

1. https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs#L174-L182
2. https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/bridges/snowbridge/pallets/inbound-queue-v2/src/lib.rs#L268

yrong replied:
Seems you only map a subset of the low-level `SendError` here? Why not  just nest the actual send error here as V1?

I prefer to add `PalletError` macro to `SendError` and remove the convert logic.  More detail in the original thread https://github.com/paritytech/polkadot-sdk/pull/6706#discussion_r1934957737

vgeddes replied:
I think `PalletError` complicates the Error type, so I changed my mind about it.

By the way, both inbound-queue-v2 and pallet-xcm both only map a subset of the low-level `SendError`. So I think we need standardize on one pattern.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/fa90934b1b9010aebc1d664a2f227850f4c57067

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -78,25 +79,38 @@ pub mod pallet {
 		type FeeAsset: Get<Location>;
 
-		type DeliveryFee: Get<Asset>;
+		/// RemoteExecutionFee for the execution cost on bridge hub
+		type RemoteExecutionFee: Get<Asset>;
```
Actually, I realized this hardcoding of fees on-chain is an anti-pattern. One of the major reasons for developing V2 was to move fee calculations offchain. So instead, `create_agent` and `register_token` extrinsics should take an extra fee parameter for bridgehub.

yrong replied:
> create_agent and register_token extrinsics should take an extra fee parameter for bridgehub.

I'm not sure this is nessessary. I treat the `RemoteExecutionFee` more like a mechanism to avoid spamming. Same as the base fee in V2 exporter.
https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/cumulus/parachains/runtimes/assets/asset-hub-westend/src/xcm_config.rs#L667

All the fees are actually covered by `fee` in ether. Including the execution cost on BH. It's the UI/relayer's responsibility to sum all the fees including and convert to ether to pay.

vgeddes replied:
> I'm not sure this is nessessary. I treat the RemoteExecutionFee more like a mechanism to avoid spamming.

This doesn't make sense. AH-SystemFrontend is sending an XCM to BH, and that the execution of that XCM requires fees to be paid on BH. You're even including `PaysFee(T::RemoteExecutionFee)` in the instruction.

> All the fees are actually covered by fee in ether. Including the execution cost on BH. It's the UI/relayer's responsibility to sum all the fees including and convert to ether to pay.

This does not solve the problem of how to pay for XCM execution on BH. And I don't want to hardcode that fee in the runtime.

yrong replied:
Make sense, I'll make change accordingly.

vgeddes replied:
Actually, I think we need to discuss this BH fee issue with Adrian next week. Maybe its too complicated for the user to figure out what fee to pass for BH. It is also problematic for governance origins, which need free execution.

In the meantime what about adding an `UnpaidExecution` in the message and modify XCM executor config in bridgehub-westend runtime to accept unpaid execution from SystemFrontendLocation.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/63e3e6674b75c264e85ae3eab69db280ce5056e0

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,64 @@ pub mod pallet {
 	impl<T: Config> Pallet<T> {
 		/// Call `create_agent` to instantiate a new agent contract representing `origin`.
-		/// - `origin`: Can be any sovereign `Location`
 		/// - `fee`: Fee in Ether paying for the execution cost on Ethreum
 		#[pallet::call_index(1)]
 		#[pallet::weight(T::WeightInfo::create_agent())]
 		pub fn create_agent(origin: OriginFor<T>, fee: u128) -> DispatchResult {
```
See my comment above, we shouldn't be hardcoding the bridgehub fee.

```suggestion
		/// - `relayer_fee`: Relayer fee in Ether for relaying message to Ethereum and dispatching it.
		/// - `acceptance_fee`: Fee in native relay token for accepting the order for delivery
		#[pallet::call_index(1)]
		#[pallet::weight(T::WeightInfo::create_agent())]
		pub fn create_agent(origin: OriginFor<T>, relayer_fee: u128, acceptance_fee: u128) -> DispatchResult {

```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
-			T::AssetTransactor::withdraw_asset(&T::DeliveryFee::get(), &origin_location, None)
+			T::AssetTransactor::check_out(origin, fee, &dummy_context);
+
```
unnecessary whitespace
```suggestion
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
+
+			T::AssetTransactor::withdraw_asset(fee, origin, None)
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
+
```
unnecessary whitespace
```suggestion
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
 
+			let topic = blake2_256(&("SnowbridgeFrontend", call.clone()).encode());
+
 			let xcm: Xcm<()> = vec![
```
As with the inbound-queue, you should descend the origin to the pallet of the `system-frontend` pallet on AH.

https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/bridges/snowbridge/primitives/inbound-queue/src/v2/converter.rs#L200

https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/bridges/snowbridge/pallets/inbound-queue-v2/src/mock.rs#L172

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -79,20 +78,19 @@ pub mod pallet {
 		type OutboundQueue: SendMessage;
 
-		/// Origin check for XCM locations that can create agents
+		/// Origin check for XCM locations that transact with this pallet
 		type SiblingOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
```
```suggestion
		/// Origin check for locations acting as a frontend for this pallet
		type FrontendOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
```

I added another comment related to this. You should rename this to `FrontendOrigin` and make sure the implementation of `FrontendOrigin` checks that the location is like `/Para(AssetHub)/PalletIndex(SystemFrontend)`.

yrong replied:
Just curious is this nessessary?  Currently `SiblingOrigin`  is configured to accept only the root location of AH(1, Parachain(1000)) which I'd assume can't be manipulated by user origin.

vgeddes replied:
We do this because its good practice to be specific and restrictive when authorizing calls. The root origin `AH(1, Parachain(1000))` doesn't give us any indication of what pallet on AH is making the call.

In any case, this is what we in the inbound-queue for BH->AH xcms. Parity asked us to do it. And they are going to hit us with same feedback unless we do it now.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/15f96644b1318f28927943aedfdc1280aeb8af5d

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -199,10 +190,9 @@ pub mod pallet {
-				Ok(RawOrigin::Root) => Ok(Here.into()),
-				_ => T::SiblingOrigin::ensure_origin(origin),
-			}?;
+			let origin_location = T::SiblingOrigin::ensure_origin(origin)?;
```
This doesn't make sense. The `origin` in this context is the system-frontend pallet on AH.

You need to add an additional `location` parameter` which is original origin, passed by the system-frontend pallet when calling this function

vgeddes commented on bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs:
```diff
@@ -292,7 +295,6 @@ pub mod pallet {
 			let order = <PendingOrders<T>>::get(nonce).ok_or(Error::<T>::InvalidPendingNonce)?;
 
-			// No fee for governance order
 			if !order.fee.is_zero() {
```
Easier to read:
```suggestion
			if order.fee > 0 {
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,64 @@ pub mod pallet {
 
+			let topic = blake2_256(&("SnowbridgeFrontend", call.clone()).encode());
+
 			let xcm: Xcm<()> = vec![
```
See https://github.com/vgeddes/polkadot-sdk/pull/6/files#r1954400482. Need to prepend `DescendOrigin([PalletIndex(SystemFrontend)])`

vgeddes replied:
Both the `create_agent` and `register_token` dispatchables are duplicating code to generate almost the same XCM message. This should be factored out into a common function, for example:

```rust
fn make_transact_xcm(call: BridgeHubRuntime>) -> Xcm<()>
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +119,64 @@ pub mod pallet {
 					fallback_max_weight: None,
 				},
+				ExpectTransactStatus(MaybeErrorCode::Success),
+				SetTopic(topic),
```
We should not let the leftover execution fee be trapped. The origin user won't be able to easily claim it on BH.

Easiest to just transfer the leftover fee to the sovereign account of the `/Para(AssetHub)/PalletIndex(SystemFrontend)`

```suggestion
				DepositAsset {
					assets: Wild(AllOf { id: message.execution_fee.id, fun: WildFungible }),
					beneficiary: SELF,
				}
				SetTopic(topic),
```

And note that `ExpectTransactStatus` will cause the xcm execution to fail before reaching `DepositAsset`.

yrong replied:
Since we use `Unpaid` now, there is no need to `DepositAsset`

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
 					fallback_max_weight: None,
 				},
+				ExpectTransactStatus(MaybeErrorCode::Success),
+				SetTopic(topic),
```
ditto, need to handle leftover fees.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
 			let asset_location: Location =
 				(*asset_id).try_into().map_err(|_| Error::<T>::UnsupportedLocationVersion)?;
 
-			let mut checked = false;
-			if asset_location.eq(&origin_location) || asset_location.starts_with(&origin_location) {
-				checked = true
-			}
-			ensure!(checked, <Error<T>>::OwnerCheck);
```
Why are you removing this check? Its important to keep.

However, we should move most of check code to the runtime configuration, in the form of a [polkadot_sdk_frame::traits::ContainsPair](https://paritytech.github.io/polkadot-sdk/master/polkadot_sdk_frame/traits/trait.ContainsPair.html#) implementation.

Then this pallet can just check for authorization using `T::AssetOwner::contains(origin_location, asset_location)`

yrong replied:
https://github.com/paritytech/polkadot-sdk/pull/7402#discussion_r1949531155

We allow end user on AH  to create `DOT` or `PAL`(the native token on penpal).

Please remind me why we initially add this check, does it add more security?

vgeddes replied:
The check is very important as only the asset owner can be trusted to specify the `metadata` parameter.

Otherwise any user can register DOT with misleading or spam metadata, causing an ERC20 to be instantiated with wrong metadata. That could be extremely harmful.

yrong replied:
Does that imply you still prefer to register `DOT` from sudo origin?  Then what about the native asset on AH like [USDT](https://assethub-polkadot.subscan.io/assets/1984), who will create it?

I prefer to allow end user to create any asset. that's what `CreateOrigin`  is configured for `USDT` or any native asset.

https://github.com/paritytech/polkadot-sdk/blob/d11400478dddb6bb78d294ced9d688712343fde0/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L273

> user can register DOT with misleading or spam metadata

The economic incentive for doing this doesn’t seem apparent to me.  Or how about enforce a `AssetDeposit` as `Assets` pallets does?
https://github.com/paritytech/polkadot-sdk/blob/d11400478dddb6bb78d294ced9d688712343fde0/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L275

vgeddes replied:
> The economic incentive for doing this doesn’t seem apparent to me.

DOT, as identified by its location, can only be registered as a PNA once. So if a malicious user is the first to register DOT with false metadata, then they've effectively broken the PNA feature

No one on Uniswap on Ethereum is going to buy ERC20 DOT if it has weird or misleading metadata.

vgeddes replied:
> Does that imply you still prefer to register DOT from sudo origin? 

We may not need `force_register_token`, since `T::AssetOwner` will check if the registration is allowed for a specific (origin, asset) pair.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/f2570120b47564b5eb74eb45926f9b6a16d94445

Though one test failed to create `DOT`  which is unclear to me, so I remove it. others are all good.

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,51 +193,68 @@ pub mod pallet {
 				fee,
 			});
 
+			let topic = blake2_256(&("SnowbridgeFrontend", call.clone()).encode());
```
```suggestion
```

ditto, its not necessary to add our own topic.

yrong committed:
Fix pallet index

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -228,7 +218,7 @@ pub mod pallet {
 				decimals: metadata.decimals,
 			};
-			Self::send(origin_location, command, fee)?;
+			Self::send(origin, command, fee)?;
```
Yes this seems good. Then in the gateway contract we can authorize this origin (/AH/SystemFrontend) to register tokens.

yrong replied:
In V1 it's always `Gateway` to create the token. Does that imply you want to change the behavior to introduce agent for PNA in V2?

Would that cause trouble for the incompatible?

vgeddes replied:
`message.origin` is not an agent, its just the hash of the origin location.

What I am saying is the Gateway could just check `message.origin == hash(SystemFrontendLocation)` instead of `message.origin == hash(BridgeHubLocation)`

But this is just a thought. I need to think about it more, so I don't need you to implement anything

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -141,40 +139,36 @@ pub mod pallet {
-				Ok(RawOrigin::Root) => Ok(Here.into()),
-				_ => T::SiblingOrigin::ensure_origin(origin),
-			}?;
+			T::SiblingOrigin::ensure_origin(origin)?;
```
ditto, see comment above:

```suggestion
			T::FrontendOrigin::ensure_origin(origin)?;
```

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -141,40 +139,36 @@ pub mod pallet {
-			let location = origin_location
+			let ethereum_location = T::EthereumLocation::get();
+			let reanchored_location = origin_location
 				.clone()
 				.reanchored(&ethereum_location, &T::UniversalLocation::get())
 				.map_err(|_| Error::<T>::LocationConversionFailed)?;
 
-			let agent_id = agent_id_of::<T>(&location)?;
+			let agent_id = agent_id_of::<T>(&reanchored_location)?;
```
move to a helper function please, this code is going to end up duplicated.

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -141,40 +139,36 @@ pub mod pallet {
-			}?;
+			T::SiblingOrigin::ensure_origin(origin)?;
 
 			let origin_location: Location =
 				(*location).try_into().map_err(|_| Error::<T>::UnsupportedLocationVersion)?;
```
This is confusing naming, since reviewers will assume `origin_location` corresponds with `origin` rather than `location`.

Rather just rebind `location`:
```suggestion
			let location: Location =
				(*location).try_into().map_err(|_| Error::<T>::UnsupportedLocationVersion)?;
```

yrong replied:
I search the code base in `polkadot-sdk` there are a lot of the usage of `origin_location` which refers to location without confusion.

So I prefer to not change.

vgeddes replied:
But the initial `location` parameter is different to `origin: OriginFor`.  Reviewers will think `origin_location` is derived from `origin` rather than `location`. This is a source of confusion.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/493d24312a29518e9adc8b1c3c843fbdbbd5bef9

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -141,40 +139,36 @@ pub mod pallet {
 			// reanchor to Ethereum context
-			let location = origin_location
+			let ethereum_location = T::EthereumLocation::get();
+			let reanchored_location = origin_location
 				.clone()
 				.reanchored(&ethereum_location, &T::UniversalLocation::get())
 				.map_err(|_| Error::<T>::LocationConversionFailed)?;
 
-			let agent_id = agent_id_of::<T>(&location)?;
+			let agent_id = agent_id_of::<T>(&reanchored_location)?;
```
I think you should move this location->agent conversion to a common helper function. I can already see that this code is duplicated between create_agent and register_token. Yet the implementations differ slightly.

yrong replied:
It’s just one line of code, so I prefer to inline it. 
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/da0ef5d17a6c1a0c57bcacde6840c8205725185e

vgeddes replied:
Its not just the agent_id conversion - it includes the the reanchoring too.

There should be a single source of truth for reanchoring and conversion.

yrong replied:
> not just the agent_id conversion - it includes the the reanchoring too.

Yeah, it's covered in other commits.

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -199,10 +190,9 @@ pub mod pallet {
-				_ => T::SiblingOrigin::ensure_origin(origin),
-			}?;
+			let origin_location = T::SiblingOrigin::ensure_origin(origin)?;
+			let origin = AgentIdOf::convert_location(&origin_location)
```
Why are you not reanchoring to be relative to Ethereum like in `create_agent`?

yrong replied:
As mentioned before the `SiblingOrigin` here is root location of AH (1,X1[Parachain(1000)]), after reanchoring it will be (1,X2[GlobalConsensus(Polkadot),Parachain(1000)] which is incompatible with the agent for AH in V1.

It's fine because currently in Gateway contract we ignore the origin when register PNA. But maybe we need to add a Pallet instance as you suggested in https://github.com/vgeddes/polkadot-sdk/pull/6#discussion_r1954405622 and do the reanchoring logic consistently as other operations.

vgeddes replied:
> after reanchoring it will be (1,X2[GlobalConsensus(Polkadot),Parachain(1000)] which is incompatible with the agent for AH in V1.

Right, that is true. we need to factor out this "reanchoring and converting location -> hash" logic to a helper function. This helper function should have a compatibility match pattern that detects AH location, and uses V1 reanchoring instead of V2 reanchoring.

We can't duplicate this reanchoring/hashing logic in each governance method, it will introduce bugs.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/dbce06cf64ef54077ea725bac01fb490204b2ec8

Btw: since we introce `Descend(FrontendPallet)`, there is no chance conflict with V1.

yrong committed:
Fix for ether

yrong committed:
Remove topic id

yrong committed:
Rename to EthereumLocation

yrong committed:
Remove whitespace

yrong committed:
Update bridges/snowbridge/pallets/outbound-queue-v2/src/lib.rs

Co-authored-by: Vincent Geddes <117534+vgeddes@users.noreply.github.com>

yrong committed:
Merge pull request #13 from yrong/ron/remove-runtime-api

Remove dry run api

yrong committed:
Fix the lookup

yrong committed:
More integration tests

yrong committed:
Cleanup

yrong committed:
Improve the error types

yrong committed:
Switch to UnpaidExecution

yrong committed:
Rename to FrontendOrigin

yrong committed:
Remove function agent_id

yrong committed:
Factor out common function

vgeddes commented on bridges/snowbridge/primitives/outbound-queue/src/v2/delivery_receipt.rs:
```diff
@@ -13,7 +13,7 @@ sol! {
 
 /// Envelope of the delivery proof
```
```suggestion
/// Delivery receipt
```

vgeddes commented on bridges/snowbridge/primitives/outbound-queue/src/v2/converter/convert.rs:
```diff
@@ -109,18 +111,105 @@ where
+	/// e. Currently, `AliasOrigin` is always required, can distinguish the V2 process from V1.
+	/// 	it's required also for dispatching transact from that specific origin.
+	/// f. SetTopic is required for tracing the message all the way along.
+	fn to_ethereum_message(&mut self) -> Result<Message, XcmConverterError> {
```
Just call this `convert`

```suggestion
	fn convert(&mut self) -> Result<Message, XcmConverterError> {
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/acbb04bbed665ecca0c6a6fbb21ed4f1c8d93cba

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,52 +184,66 @@ pub mod pallet {
+						Error::<T>::from(err)
+					},
+				)?;
+			T::XcmExecutor::charge_fees(origin, price).map_err(|_| Error::<T>::FundsUnavailable)?;
```
Should mirror inbound-queue:

```suggestion
			T::XcmExecutor::charge_fees(origin, price).map_err(|_| Error::<T>::FeesNotMet)?;
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/00f5ac8081c229591ba3dcb3a85795e343c37904

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,52 +184,66 @@ pub mod pallet {
+				},
+			])
+		}
+
+		fn reanchor(location: &Location) -> Result<Location, Error<T>> {
```
```suggestion
		/// Reanchors `location` relative to BridgeHub.
		fn reanchor(location: &Location) -> Result<Location, Error<T>> {
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -164,52 +184,66 @@ pub mod pallet {
 				.map_err(|_| Error::<T>::FundsUnavailable)?;
 			Ok(())
 		}
+
+		fn build_xcm(call: Vec<u8>) -> Xcm<()> {
```
We shouldn't pass raw bytes around unless required. Its useful to preserve type information for as long as we can. Consider constraining `call` as follows:

```suggestion
		/// Build an XCM to invoke `call` on the destination parachain
		fn build_xcm(call: &impl Encode) -> Xcm<()> {
```

Or alternatively just pass `call: BridgeHubRuntime`

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/263abab30af616000d82065262078068e878975f

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +120,54 @@ pub mod pallet {
+	impl<T: Config> From<SendError> for Error<T> {
+		fn from(e: SendError) -> Self {
+			match e {
+				SendError::Fees => Error::<T>::FundsUnavailable,
```
Mirror inbound-queue error types:

```suggestion
				SendError::Fees => Error::<T>::FeesNotMet,
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +120,54 @@ pub mod pallet {
+			match e {
+				SendError::Fees => Error::<T>::FundsUnavailable,
+				SendError::NotApplicable => Error::<T>::Unreachable,
+				_ => Error::<T>::Send,
```
Mirror inbound-queue-v2 error types:

```suggestion
				_ => Error::<T>::SendFailure,
```

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/04eac741164ea28a7d76badd171fe6e4335e3dec

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -260,6 +231,20 @@ pub mod pallet {
+				.map_err(|_| Error::<T>::LocationReanchorFailed)
+		}
+
+		fn location_to_agent_id(location: &Location) -> Result<AgentId, Error<T>> {
+			let reanchored_location = Self::reanchor(location)?;
+			AgentIdOf::convert_location(&reanchored_location)
+				.ok_or(Error::<T>::LocationConversionFailed)
+		}
```
An `Agent` is an ethereum-specific construct. Within this pallet, `message.origin` is just a hash of a reanchored location.

Otherwise you introduce confusion in the sense that reviewers expect that there is an `Agent` contract on Ethereum for this location, which is not always the case.

```suggestion
		fn location_to_message_origin(location: &Location) -> Result<H256, Error<T>> {
			let reanchored_location = Self::reanchor(location)?;
			AgentIdOf::convert_location(&reanchored_location)
				.ok_or(Error::<T>::LocationConversionFailed)
		}
```

vgeddes replied:
Also see my comment above where I suggest that `T::AgentIdOf` (suggest to rename to `T::LocationHashOf`) also does reanchoring logic.

yrong replied:
https://github.com/vgeddes/polkadot-sdk/pull/6/commits/a7052509b6b5747fb1b9601c7e3a92f80b66cfff

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -240,12 +214,9 @@ pub mod pallet {
-		fn send(origin_location: Location, command: Command, fee: u128) -> DispatchResult {
-			let origin = agent_id_of::<T>(&origin_location)?;
-
+		/// Send `command` to the Gateway from a specific origin/agent
+		fn send(origin: AgentId, command: Command, fee: u128) -> DispatchResult {
```
Not all polkadot origins have an agent on Ethereum. I prefer we not use the word `Agent` in this pallet. Within the context of the pallet, the origin location is just being reanchored and hashed to a `H256`.

```suggestion
		/// Send `command` to the Gateway
		fn send(origin: H256, command: Command, fee: u128) -> DispatchResult {
```

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -79,20 +74,19 @@ pub mod pallet {
+		/// Origin check for XCM locations that transact with this pallet
+		type FrontendOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
 
 		/// Converts Location to AgentId
 		type AgentIdOf: ConvertLocation<AgentId>;
```
Is there any reason why this type can't do reanchoring as well?

Also as I suggested in other comments, the word `Agent` is confusing. An agent is an ethereum contract. Within the context of this pallet, we are really just producing an `H256` to be able to set `message.origin`.

Whether the Gateway associates that with an Agent is not something this pallet should be concerned about.

```suggestion
		/// Converts Location to AgentId
		type LocationHashOf: ConvertLocation<H256>;
```

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -141,40 +136,29 @@ pub mod pallet {
 			let command = Command::CreateAgent {};
 
-			Self::send(origin_location.clone(), command, fee)?;
+			Self::send(agent_id, command, fee)?;
```
A hashed origin location may or may not correspond to the ID of an Agent contract on ethereum.

```suggestion
			Self::send(message_origin, command, fee)?;
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +120,54 @@ pub mod pallet {
 		/// Send xcm message failure
 		Send,
 		/// Withdraw fee asset failure
 		FundsUnavailable,
```
Mirror inbound-queue-v2 error types
```suggestion
		FeesNotMet,
```

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -105,49 +120,54 @@ pub mod pallet {
-		OwnerCheck,
+		InvalidAssetOwner,
 		/// Send xcm message failure
 		Send,
```
Mirror inbound-queue-v2 error types

```suggestion
		SendFailure,
```

vgeddes commented on bridges/snowbridge/pallets/system-v2/src/lib.rs:
```diff
@@ -141,40 +136,29 @@ pub mod pallet {
-				.map_err(|_| Error::<T>::LocationConversionFailed)?;
-
-			let agent_id = agent_id_of::<T>(&location)?;
+			let agent_id = Self::location_to_agent_id(&origin_location)?;
```
```suggestion
			let message_origin = Self::location_to_message_origin(&origin_location)?;
```

yrong committed:
Revamp with EnsureOriginWithArg

vgeddes commented on bridges/snowbridge/pallets/system-frontend/src/lib.rs:
```diff
@@ -71,9 +74,12 @@ pub mod pallet {
-		type RegisterTokenOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
-
+		/// Origin check for XCM locations that can register token
+		type RegisterTokenOrigin: EnsureOriginWithArg<
```
This is too complicated. Why not the `AssetOwner: ContainsPair<Location, Location>` approach?

I don't think Parity will like the `EnsureOriginWithArg` approach. Yes, its very smart, but at the expense of clarity. It overloads the semantics of origin checking to perform business logic that should be done in the pallet. I can tell you with good certainty that they will reject it.

Its easier for reviewers if they can see in the pallet, sequentially that:
1. The origin is acceptable on its own without any other context
2. Furthermore, that the origin is allowed to register a specified asset location

yrong replied:
Just runtime config wrapped on top of `ContainsPair<Location, Location>` 

This will check whether the origin is allowed to register a specific asset.
```
let origin_location = T::RegisterTokenOrigin::ensure_origin(origin, &asset_location)?;
```

vgeddes replied:
Your approach looks smart, but I think it obscures the business logic that should be done within the pallet rather than the runtime config.

Parity won't accept it, I can tell with you with 100% certainty. Please rather add a separate check using `AssetOwner: ContainsPair<Location, Location>`

vgeddes replied:
Specifically:

```rust
origin_location = T::RegisterTokenOrigin::ensure_origin(origin)?;
ensure!(T::AssetOwner::contains(origin_location, location))?;
```

vgeddes replied:
Also the other reason why your approach is not good is because we don't need separate `T::CreateAgentOrigin` and `T::RegisterToken` origin checkers, since both `create_agent` and `register_token` accept any XCM origin, before applying additional origin verification logic.

yrong replied:
Actually I just borrowed that from `ForeignAssets` pallet, e.g. check only the bridge location can create ERC20.

https://github.com/vgeddes/polkadot-sdk/blob/19f7f03635861fc31c6f5a9b2123585e44a2153a/cumulus/parachains/runtimes/assets/asset-hub-westend/src/lib.rs#L531

I'd prefer to not reinvent the wheel.

vgeddes replied:
ok, sorry, I didn't see that Parity eventually moved to this approach too.

yrong committed:
Update bridges/snowbridge/primitives/outbound-queue/src/v2/delivery_receipt.rs

Co-authored-by: Vincent Geddes <117534+vgeddes@users.noreply.github.com>

yrong committed:
Rename to location

yrong committed:
Merge branch 'ron/snowbridge-v2-for-comments' of https://github.com/yrong/polkadot-sdk into ron/snowbridge-v2-for-comments

yrong committed:
Rename to convert

yrong committed:
Rename to FeesNotMet

yrong committed:
Update bridges/snowbridge/pallets/system-frontend/src/lib.rs

Co-authored-by: Vincent Geddes <117534+vgeddes@users.noreply.github.com>

yrong committed:
Constraining call as Encode

yrong committed:
Rename to SendFailure

yrong committed:
Polish system-v2

yrong committed:
Fix compile

yrong committed:
Fix build xcm with barrier config

yrong committed:
Couple with system-v1
